Libraries/LibGC/Heap+Swift.swift

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2024, Andrew Kaster <andrew@ladybird.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+import AK
+@_exported import GCCxx
+
+extension GC.Heap {
+    public func withDeferredGC<R, E>(_ body: () throws(E) -> R) throws(E) -> R {
+        self.defer_gc()
+        defer { self.undefer_gc() }
+        return try body()
+    }
+}
+
+// FIXME: Cell and Cell::Visitor are not imported properly, so we have to treat them as OpaquePointer
+public protocol HeapAllocatable {
+    static func allocate(on heap: GC.Heap) -> UnsafeMutablePointer<Self>
+
+    init(cell: OpaquePointer)
+
+    func finalize()
+    func visitEdges(_ visitor: OpaquePointer)
+
+    var cell: OpaquePointer { get }
+}
+
+// FIXME: Figure out why other modules can't conform to HeapAllocatable
+public struct HeapString: HeapAllocatable {
+    public var string: Swift.String
+
+    public init(cell: OpaquePointer) {
+        string = "Hello, World!"
+        self.cell = cell
+    }
+
+    // FIXME: HeapAllocatable cannot be exposed to C++ yet, so we're off to void* paradise
+    public static func create(on heap: GC.Heap) -> OpaquePointer {
+        precondition(heap.is_gc_deferred())
```

Comment:
but _why_

Reply:
I ran into crashes if the entire init sequence was not gc-deferred. If I stored the swift type pointer on the stack without having the Cell pointer either on the stack or stashed in a GCPtr, it risks being GCd if we collect immediately.

Reply:
Maybe worth a comment then?